### PR TITLE
[TODO: I need to use a better API name] Add a custom print HOP to be able to print in both fwd, bwd, and take in custom printing functions

### PR DESCRIPTION
**Pull request by yushangdi**
This PR introduces a new Higher Order Operator (HOP) for observing tensors and gradients 
  during forward and backward passes in torch.compile regions. The HOP uses opaque objects 
  to wrap callback functions, enabling compile-safe execution without a global registry.   
                                                            

 ## Key Implementation Details

  1. Opaque Object Pattern: Callbacks are wrapped in CallbackWrapper(OpaqueBase) and
  registered with register_opaque_type(CallbackWrapper, typ="reference"). This avoids a
  global registry and integrates with dynamo's opaque object tracing.
  2. Two HOPs:
    - compile_print_fwd(fwd_callback, bwd_callback, *tensors) -> None: Forward pass
    - compile_print_bwd(bwd_callback, *grads) -> None: Backward pass
  3. Pure Side-Effect: Both HOPs return None and are registered as _EffectType.ORDERED to
  prevent DCE.
  4. Backward via Tensor Hooks: The autograd implementation uses register_hook on input
  tensors to fire backward callbacks even when the HOP's output isn't used in the
  computation graph.
  5. FakeScriptObject Handling: During compiled execution, CallbackWrapper becomes a
  FakeScriptObject. The _get_real_callback() helper extracts the real callback via
  fake_obj.real_obj.
  6. Schema Generation: Updated HopSchemaGenerator.add_arg() to recognize opaque types
  before checking if arguments are callable, preventing false "Expect callable to be a
  GraphModule" errors.

 ## Files Changed

  - torch/_higher_order_ops/new_compile_print.py - Main implementation with CallbackWrapper
   opaque object
  - torch/_higher_order_ops/effects.py - Register HOPs as effectful
  - torch/_higher_order_ops/schema.py - Handle opaque types in schema generation
  - torch/_dynamo/variables/higher_order_ops.py - Dynamo variable classes for tracing
  - test/higher_order_ops/test_new_compile_print.py - 15 tests covering eager, compiled,
  make_fx, aot_function, graph structure, and callback isolation

##  Test Coverage

  - Eager and compiled execution
  - make_fx and aot_function tracing
  - Callbacks with arbitrary tensor computations (verified they don't affect gradients)
  - Side-effect-only usage (verified HOPs aren't DCE'd)
  - Graph structure verification with assertExpectedInline

## Usage example:

```python


        def fwd_f(t):
            print(f"Forward: shape={t.shape}, mean={t.mean().item():.4f}")

        def bwd_f(t):
            print(f"Backward: shape={t.shape}, mean={t.mean().item():.4f}")

        cp = make_compile_print(fwd_f, bwd_f)

        def fn(x, y):
            # Call compile_print but don't use its output
            cp(x, y)
            # Use the original tensors for the computation
            return x.sum() + y.sum()

        x = torch.randn(3, 3, requires_grad=True)
        y = torch.randn(3, 3, requires_grad=True)

# with compile
            compiled_fn = torch.compile(fn, backend="aot_eager", fullgraph=True)
            loss = compiled_fn(x, y)
            loss.backward()


# with make_fx

        def nop_compiler(gm, _):
            return make_boxed_func(gm)

        aot_fn = aot_function(fn, fw_compiler=nop_compiler, bw_compiler=nop_compiler)

        with patch("builtins.print") as mock_print:
            loss = aot_fn(x, y)
            loss.backward()
```

  ---
  This PR was authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

**Comment by yushangdi** [comment]
waiting for https://github.com/pytorch/pytorch/pull/174122 to prevent it from being DCE'd

File: torch/_higher_order_ops/new_compile_print.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+    Returns:
+        Tuple of the input tensors, unchanged.
+    """
+    fn = make_compile_print(fwd_f, bwd_f)
+    return fn(*args)
+
+
+def _register_dynamo_hop_variables() -> None:
+    """Register the compile_print HOPs with dynamo."""
```
Comment (fxdawnn) [inline torch/_higher_order_ops/new_compile_print.py]: this might happen within dynamo itself?

File: test/higher_order_ops/test_compile_print.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        # Create the compile_print function ahead of time (before torch.compile)
+        cp = make_compile_print(fwd_f, bwd_f)
+
+        def fn(x, y):
+            out = cp(x, y)
+            return out[0].sum() + out[1].sum()
+
+        x = torch.randn(3, 3, requires_grad=True)
+        y = torch.randn(3, 3, requires_grad=True)
+
```
Comment (fxdawnn) [inline test/higher_order_ops/test_compile_print.py]: can you also examine the graph here and also the io output?
